doc/how-to-freeze-versions.rst

```diff
@@ -0,0 +1,247 @@
+private team.  If anyone using it can somehow have different inputs to
+the solver then the ``.freeze`` file can be troublesome.  It can prevent the
+solver from finding a different version of a dependency that would satisfy a
+different architecture or a different compiler version and boot libraries.
+
+.. Warning::
+
+    If publishing a package to Hackage, not matter what kind of component it
+    contains, don't include a ``.freeze`` file, don't add it to any field of the
+    package description that would have ``cabal sdist`` include it in the
+    ``.tar.gz``. In general, don't include anything in the package description
+    that relates to the project environment, like ``cabal.project`` or
+    ``cabal.project.local``.
+
+Freezing workflows
+^^^^^^^^^^^^^^^^^^
+
+.. Warning::
+    For each of these workflows, you may have to first delete the
+    ``index-state`` line from ``cabal.project`` (and from
+    ``cabal.project.freeze`` if it exists) and then run ``cabal update`` to
+    ensure that cabal will have newer versions to re-resolve the dependencies
+    with. Alternatively, you can run ``cabal update
+    --ignore-project``.
+
+Freeze
+    If the ``cabal.project.freeze`` file doesn't exist, generating one is a
+    great way to see what versions of dependencies are currently being used even
+    if you choose to discard the ``.freeze`` file after inspecting it.
+
+Thaw, Freeze
+    If you changed the version ranges of any of the dependencies in any of your
+    project's package descriptions, in any ``.cabal`` file, then delete the
+    ``cabal.project.freeze`` file if it already exists and run ``cabal freeze``
+    to generate fresh version of ``cabal.project.freeze``.
+
+Freeze, Freeze (Freezing Harder)
+    You might in some cases want to skip deletion of ``cabal.project.freeze``,
+    but keep in mind that in that case ``cabal freeze`` will use existing
+    ``cabal.project.freeze`` when resolving dependencies, therefore not updating
+    any existing dependencies, only adding new ones.
+
+Partial Thaw, Freeze
+    If you want to a pick up a different version of a single dependency, you can
+    delete its constraint from ``cabal.project.freeze`` and then run ``cabal
+    freeze`` again.
+
+.. Note::
+
+    If not sure, pick the "thaw, freeze" workflow, as it is the safest, the
+    simplest and the most common. Finally, you will always want to commit the
+    changed ``cabal.project.freeze`` to version control.
+
+Ensuring everything is frozen
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. Note::
+
+    If the ``.freeze`` file already has version equality constraints for every
+    package that is a dependency of the project, then the solver will not be
+    able to find a different version for any of them, the ``.freeze`` file
+    cannot change and, at that point when every dependency is frozen, ``cabal
+    freeze`` becomes an idempotent operation.
+
+Adding a dependency to one of the packages in a project without freezing harder
```

Comment:
This line doesn't make sense unless you read the previous section to know what "freeze harder" means. Some readers might land on this section by following a link to it directly. Perhaps be more explicit?
```suggestion
Adding a dependency to one of the packages in a project without running ``cabal freeze`` again
```
or
```suggestion
Adding a dependency to one of the packages in a project without regenerating the ``cabal.project.freeze`` file using one of the methods described above
```
(Perhaps with a link to the above section)